pkg/config/structure/unmarshal_test.go

```diff
@@ -148,19 +180,607 @@ type FeatureConfig struct {
-	confYaml := `
+func TestUnmarshalKeyAsBool(t *testing.T) {
+	testcases := []struct {
+		name string
+		conf string
+		want bool
+		skip bool
+	}{
+		{
+			name: "string value to true",
+			conf: `
 feature:
   enabled: "true"
-`
-	mockConfig := mock.NewFromYAML(t, confYaml)
-	mockConfig.SetKnown("feature")
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "yaml boolean value true",
+			conf: `
+feature:
+  enabled: true
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "string value to false",
+			conf: `
+feature:
+  enabled: "false"
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "yaml boolean value false",
+			conf: `
+feature:
+  enabled: false
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "missing value is false",
+			conf: `
+feature:
+  not_enabled: "the missing key should be false"
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "string 'y' value is true",
+			conf: `
+feature:
+  enabled: y
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "string 'yes' value is true",
+			conf: `
+feature:
+  enabled: yes
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "string 'on' value is true",
+			conf: `
+feature:
+  enabled: on
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "string '1' value is true",
+			conf: `
+feature:
+  enabled: "1"
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "int 1 value is true",
+			conf: `
+feature:
+  enabled: 1
+`,
+			want: true,
+			skip: false,
+		},
+		{
+			name: "string 'n' value is false",
+			conf: `
+feature:
+  enabled: n
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "string 'no' value is false",
+			conf: `
+feature:
+  enabled: no
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "string 'off' value is false",
+			conf: `
+feature:
+  enabled: off
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "string '0' value is false",
+			conf: `
+feature:
+  enabled: "0"
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "int 0 value is false",
+			conf: `
+feature:
+  enabled: 0
+`,
+			want: false,
+			skip: false,
+		},
+		{
+			name: "yaml empty string value false",
+			conf: `
+feature:
+  enabled: ""
+`,
+			want: false,
+			skip: true, // TODO: should this include falsey ? (nil, zero values like empty string, etc.)
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.skip {
+				t.Skip("Skipping test case")
+			}
+
+			mockConfig := mock.NewFromYAML(t, tc.conf)
+			mockConfig.SetKnown("feature")
+
+			var feature = FeatureConfig{}
+			err := UnmarshalKey(mockConfig, "feature", &feature)
+			assert.NoError(t, err, "%s failed to marshal: %s", tc.name, err)
+
+			assert.Equal(t, feature.Enabled, tc.want, "%s unexpected marshal value, want: %s got: %s", tc.name, tc.want, feature.Enabled)
+		})
+	}
+}
 
-	var feature = FeatureConfig{}
-	err := UnmarshalKey(mockConfig, "feature", &feature)
-	assert.NoError(t, err)
+type UintConfig struct {
```

Comment:
Should we make all the structs, and struct fields used in this tests private? I think is a good practice 😄 

`UintConfig`, `FloatConfig`, `StringConfig`, ...

Reply:
>  struct fields used in this tests private? I think is a good practice

reflect [will panic if the struct fields](https://pkg.go.dev/reflect#Value.Interface) are not exported

```
    --- FAIL: TestUnmarshalKeyAsString/string_value_config_map (0.00s)
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
        panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
```

I thought this also happened for the structs to be exported but not the case, fixed in https://github.com/DataDog/datadog-agent/pull/29604/commits/72eb02fb77457af24888fbad670067bb6a6dd9f7

Reply:
TIL about `UnmarshalKey` panicing if fields are not public 😄